hazelcast/src/main/java/com/hazelcast/instance/GroupProperties.java

```diff
@@ -221,6 +227,37 @@
      */
     public static final String PROP_JCACHE_PROVIDER_TYPE = "hazelcast.jcache.provider.type";
 
+    /**
+     * Result size limit for query operations on maps.
+     * <p/>
+     * This value defines the maximum number of returned elements for a single query result. If a query exceeds this number of
+     * elements a {@link QueryResultSizeExceededException} will be thrown.
+     * <p/>
+     * This feature is in place to prevent an OOME if a single node is requesting the whole data set of the cluster, e.g. by
```

Comment:
Just the local pre-check should be limited to the usage of `TruePredicate` (directly or indirectly). The query result limit (checked in `QueryOperation`) should work for all queries.